### PR TITLE
Add/coming soon add flag on create and disable it on launch

### PR DESCRIPTION
**Pull request by roo2**
Moved from D50753-code
On creating a new site, if wpcom_public_coming_soon is passed in, apply this as a site option

When changing the privacy mode of a site (e.g. when launching a site) disable public coming soon if it is enabled.

Also consolidate the naming for the wpcom_public_coming_soon flag.

### Testing instructions
Check out this branch and sync to your sandbox with `yarn dev --sync` 
Ensure `define('WPCOM_PUBLIC_COMING_SOON', true);` is in your sandbox.php
Go to the `/start/domains?flags=coming-soon-v2` endpoint (enable coming-soon-v2 flag)
Create a site
The site should have wpcom_public_coming_soon set to 1
`wp option --url=https://yourtestsite.wordpress.com get wpcom_public_coming_soon`

Create a site from the `/new?flags=coming-soon-v2` endpoint
Check `wpcom_public_coming_soon` before launching the site, it should be 1
Launch the site
`wpcom_public_coming_soon` should be set to 0

Also, with an unlaunched site, go to `/wp-admin/options-reading.php` and change it to public
The `wpcom_public_coming_soon` flag should also be updated

### CODE REVIEW

File: apps/editing-toolkit/editing-toolkit-plugin/coming-soon/coming-soon.php [hunk 1]
```diff
@@ -109,6 +109,44 @@ function add_coming_soon_page_to_new_site() {
...
+ * @param string $value     the new value of blog_public.
+ */
+function disable_coming_soon_on_privacy_change( $old_value, $value ) {
+	if ( 0 !== (int) $old_value || 0 === (int) $value ) {
+		// Do nothing if not moving from public-not-indexed.
+		return;
+	}
+	update_option( 'wpcom_public_coming_soon', 0 );
+}
+add_action( 'update_option_blog_public', __NAMESPACE__ . '\disable_coming_soon_on_privacy_change', 10, 2 );
```
Comment (lsl) [inline apps/editing-toolkit/editing-toolkit-plugin/coming-soon/coming-soon.php:127]: do we need this?

**Review by lsl — APPROVED** [review]
- [x] Check out this branch and sync to your sandbox with `yarn dev --sync`
- [x] Ensure `define('WPCOM_PUBLIC_COMING_SOON', true);` is in your `sandbox.php`
- [x] Go to the `/start/domains?flags=coming-soon-v2` endpoint (enable coming-soon-v2 flag)
- [x] Create a site
- [x] The site should have `wpcom_public_coming_soon` set to 1 `wp option --url=https://yourtestsite.wordpress.com get wpcom_public_coming_soon`
- [x] Create a site from the /new?flags=coming-soon-v2 endpoint
- [x] Check wpcom_public_coming_soon before launching the site, it should be 1
- [ ] Launch the site `wpcom_public_coming_soon` should be set to 0
- [x] Also, with an unlaunched site, go to /wp-admin/options-reading.php and change it to public
- [x] ^ also checked to private
- [x] The wpcom_public_coming_soon flag should also be updated

The only problem I found was that under v2 sites were being created in a launched state. (Or more precisely a non launched state - the setting option wasn't set, I think the logic then falls onto private/public which is now changed so we've probably broken the launch flow with this change.)

Setting launch-status site option = 'unlaunched' should fix that part.

I did this manually which got the v2 custom page to load (cc @ramonjd we need to remove/disable this, @p-jackson is this maybe what you were seeing the other day?)

I also noticed the banner that shows in v1 unlaunched / coming soon no longer shows in v2, I'm guessing that is doing the same logic check for private/public and bailing early.

Approving because as long as we don't enable v2 this shouldn't cause any problems :crossed_fingers: 

Didn't test gutenboarding. Assuming this is fine.

File: apps/editing-toolkit/editing-toolkit-plugin/coming-soon/coming-soon.php (hunk 1 above)
Comment (ramonjd) [reply apps/editing-toolkit/editing-toolkit-plugin/coming-soon/coming-soon.php:127]: > do we need this?

I think we do. It's trying to solve what I've done in D50760-code (see `launch_site_on_privacy_change()`, but in the plugin, which I think is a good place for it.

D50760-code still ensures that we're setting the right launch status for new, public + Coming Soon sites (see `wpmu_new_blog_set_launch_status()`) - is that possible to do in the plugin as well so we can bypass those freaky conditions in the backend?

**Comment by p-jackson** [comment]
@roo2 looks like this is doing what I'm doing in D50760-code, but it's in the ETK which I think is better. I think we still need some of the D50760-code changes in order to teach the v1 code not to auto-launch a site just because it's not private.

I should take the launching code out of D50760-code so we can have it in this PR instead. But what do you think about updating the launch status by hooking the `update_option_wpcom_public_coming_soon` action instead of `update_option_blog_public`? Also here p1602198382121100-slack-G01AY5RDDU0

You might still feel that hooking the `blog_public` setting is the right thing to do to catch updates that happen in wp-admin. It just seems to me that a user wouldn't expect changing an obscure setting like "let Google index my site" would launch their site.

**Comment by lsl** [comment]
> You might still feel that hooking the blog_public setting is the right thing to do to catch updates that happen in wp-admin. It just seems to me that a user wouldn't expect changing an obscure setting like "let Google index my site" would launch their site.

Agree on this, I think decoupling this from launch will decouple the coming soon / blog_public options one step further too.

**Comment by roo2** [comment]
> It just seems to me that a user wouldn't expect changing an obscure setting like "let Google index my site" would launch their site.

Hmmm that is a really good point.

But it is the existing behaviour with launch-status source/xref/trunk/wp-content/mu-plugins/launch-site.php#21. It is currently only triggered when changing from Private, to Public  though. It makes a lot more sense that that would launch your site then changing the "Google can index" setting.

hmm we could put something in source/xref/trunk/public.api/rest/wpcom-json-endpoints/class.wpcom-json-api-launch-site-endpoint.php#24 to set wpcom_public_coming_soon directly. I think that the reason they originally added the action on updating `blog_public` was that there might be more ways to update the blog_public setting and we don't want the settings to get out of sync. 

I think that conceptually, if the user sets the site to public, there site should also be "launched" and no longer "coming soon"

But overall I agree that as it stands now, it would be strange if unchecking the "Discourage search engines from indexing this site" checkbox launched your site

**Comment by lsl** [comment]
> I think that the reason they originally added the action on updating blog_public was that there might be more ways to update the blog_public setting and we don't want the settings to get out of sync.

I don't think it matters all that much which way we go here. 

Ideally I would say lets not couple the options and if they get out of sync that is ok - provided we correctly apply precedence when it comes to private > unlaunched / coming soon > launched.

**Comment by p-jackson** [comment]
> hmm we could put something in source/xref/trunk/public.api/rest/wpcom-json-endpoints/class.wpcom-json-api-launch-site-endpoint.php#24 to set wpcom_public_coming_soon directly.

Yes I think we could do this. It's probably more direct/explicit than hooking the `update_option_wpcom_public_coming_soon` action. You're right about v1 needing to catch all the cases when `blog_public` changed, and I think that was the right decision for the v1 design. I'm just don't think it's correct for v2.

**Comment by roo2** [comment]
ok, I think I have my head around this.

* I've reduced this PR to just handling the `wpcom_public_coming_soon` flag for new sites.
* Philip's D50760-code diff handles setting `launch-status` for new sites which @lsl pointed out was unset ( BTW thanks for the thorough testing @lsl )
* I've created a diff D50908-code to handle launching the site, rather than subscribing to updates on blog_public, because  blog_public gets updated when unchecking the "Discourage google from indexing my site" flag, which probably shouldn't launch your site!

So if we can get these three things merged, we should have new sites set to coming soon correctly and launching correctly

File: apps/editing-toolkit/editing-toolkit-plugin/coming-soon/coming-soon.php [hunk 2]
```diff
@@ -109,6 +109,44 @@ function add_coming_soon_page_to_new_site() {
...
+ * @param int    $blog_id    Blog ID.
+ * @param int    $user_id    User ID.
+ * @param string $domain     Site domain.
+ * @param string $path       Site path.
+ * @param int    $site_id    Site ID.
+ * @param array  $meta       Meta data. Used to set initial site options.
+ */
+function add_option_to_new_site( $blog_id, $user_id, $domain, $path, $site_id, $meta ) {
+	if ( 0 === $meta['public'] && 1 === (int) $meta['options']['wpcom_public_coming_soon'] ) {
+		add_blog_option( $blog_id, 'wpcom_public_coming_soon', 1 );
```
Comment (lsl) [inline apps/editing-toolkit/editing-toolkit-plugin/coming-soon/coming-soon.php:142]: I suspect this must also set launch-status = unlaunched.

**Comment by lsl** [comment]
Retested with D50908-code applied as well as this PR's etk built and synced.

Signup with v2 enabled sandbox define + calypso enabled flag for /start: 
- wpcom_coming_soon = not set
- wpcom_public_coming_soon = 1
- launch-status = not set
- blog_public = 0
- Settings page shows public not indexed (correct from blog_public but weird without v2), enabling v2 for the setting pages shows this as coming soon correctly so I think this is ok.
- Signed into site (sandboxed v2 enabled) - coming soon launch banner is shown
- wpcom_public_coming_soon_page_id = 28 - this needs to be fixed
- Custom coming soon page is still being shown for launched sites, I don't think this matter here but needs a fix.

Signup with v2 installed but disabled on sandbox define + disabled in flag /start:
- Basically as you would expect v1 to behave
- D50908-code is being applied regardless of feature flag state in the launch process, I think this is ok but we should be also setting v1 option to 0 here for the meantime - I'm concerned it may create problems for people who launch in v1 and then somehow manage to get their site set to private without using calypso settings page to do it (wp admin maybe? cli?)

So to summarize..

D50908-code - also cleanup v1 settings
this pr - also set launch status

atomic is still a big ???

**Comment by p-jackson** [comment]
D50760-code is deployed so `launch-status` should be set now for new `/start` sites 🤞

**Comment by p-jackson** [comment]
Sweet! Was just gonna say I just tried this with D50908-code also applied and everything actually worked end to end! 🥳
